internal/cluster/clusteradapter/node_pool_processor.go

```diff
@@ -0,0 +1,122 @@
+
+import (
+	"context"
+
+	"emperror.dev/errors"
+	"github.com/jinzhu/gorm"
+	"github.com/mitchellh/mapstructure"
+
+	"github.com/banzaicloud/pipeline/internal/cluster"
+	"github.com/banzaicloud/pipeline/internal/cluster/distribution"
+	"github.com/banzaicloud/pipeline/pkg/cluster/eks"
+	"github.com/banzaicloud/pipeline/pkg/providers"
+	"github.com/banzaicloud/pipeline/src/model"
+)
+
+// DistributionNodePoolProcessor processes a node pool request according to its own distribution.
+type DistributionNodePoolProcessor struct {
+	db *gorm.DB
+}
+
+// NewDistributionNodePoolProcessor returns a new DistributionNodePoolProcessor.
+func NewDistributionNodePoolProcessor(db *gorm.DB) DistributionNodePoolProcessor {
+	return DistributionNodePoolProcessor{
+		db: db,
+	}
+}
+
+// ProcessNew processes a new node pool descriptor.
+func (v DistributionNodePoolProcessor) ProcessNew(
+	_ context.Context,
+	c cluster.Cluster,
+	rawNodePool cluster.NewRawNodePool,
+) (cluster.NewRawNodePool, error) {
+	switch {
+	case c.Cloud == providers.Amazon && c.Distribution == "eks":
+		var nodePool distribution.NewEKSNodePool
+
+		err := mapstructure.Decode(rawNodePool, &nodePool)
+		if err != nil {
+			return rawNodePool, errors.Wrap(err, "failed to decode node pool")
+		}
+
+		var eksCluster model.EKSClusterModel
+
+		err = v.db.
+			Where(model.EKSClusterModel{ClusterID: c.ID}).
+			Preload("Subnets").
+			First(&eksCluster).Error
+		if gorm.IsRecordNotFoundError(err) {
+			return rawNodePool, errors.NewWithDetails(
+				"cluster model is inconsistent",
+				"clusterId", c.ID,
+			)
+		}
+		if err != nil {
+			return rawNodePool, errors.WrapWithDetails(
+				err, "failed to get cluster info",
+				"clusterId", c.ID,
+				"nodePoolName", nodePool.Name,
+			)
+		}
+
+		// Default node pool image
+		if nodePool.Image == "" {
+			image, err := eks.GetDefaultImageID(c.Location, eksCluster.Version)
+			if err != nil {
+				return rawNodePool, err
+			}
+
+			rawNodePool["image"] = image
+		}
+
+		// Resolve subnet ID or fallback to one
+		if nodePool.Subnet.SubnetId == "" && nodePool.Subnet.Cidr != "" && nodePool.Subnet.AvailabilityZone != "" {
+			for _, s := range eksCluster.Subnets {
+				if s.Cidr != nil && *s.Cidr == nodePool.Subnet.Cidr && s.AvailabilityZone != nil && *s.AvailabilityZone == nodePool.Subnet.AvailabilityZone {
+					rawNodePool["subnet"] = map[string]interface{}{
+						"subnetId":         *s.SubnetId,
+						"cidr":             nodePool.Subnet.Cidr,
+						"availabilityZone": nodePool.Subnet.AvailabilityZone,
+					}
+				}
+			}
+		} else if nodePool.Subnet.SubnetId == "" {
+			// TODO: is this necessary?
+			if len(eksCluster.Subnets) == 0 {
+				return rawNodePool, errors.New("cannot resolve subnet")
+			}
+
+			rawNodePool["subnet"] = map[string]interface{}{
+				"subnetId":         *eksCluster.Subnets[0].SubnetId,
+				"cidr":             *eksCluster.Subnets[0].Cidr,
+				"availabilityZone": *eksCluster.Subnets[0].AvailabilityZone,
+			}
+		}
```

Comment:
if `nodePool.Subnet.SubnetId != "" ` than we should validate that the provided subnet id is contained by `eksCluster.Subnets`

Reply:
That's done in the validator: https://github.com/banzaicloud/pipeline/pull/2617/files#diff-001ac5a7d84bdf4c47843ca75f19cf2dR89